discord.go

```diff
@@ -0,0 +1,266 @@
+			n := 64
+			messages := make([]brain.Message, n)
+			n, _, err := robo.brain.Recall(ctx, ch.Learn, "", messages)
+			if err != nil {
+				_ = session.InteractionRespond(event.Interaction, &discordgo.InteractionResponse{
+					Type: discordgo.InteractionResponseChannelMessageWithSource,
+					Data: &discordgo.InteractionResponseData{
+						Flags:           discordgo.MessageFlagsEphemeral,
+						Content:         err.Error(),
+						AllowedMentions: nil,
+					},
+				})
+				return
+			}
+			deleted := 0
+			for _, message := range messages {
+				if !strings.Contains(message.Text, prompt) {
+					continue
+				}
+				_ = robo.brain.Forget(ctx, ch.Learn, message.ID)
+				deleted++
+			}
+			_ = session.InteractionRespond(event.Interaction, &discordgo.InteractionResponse{
+				Type: discordgo.InteractionResponseChannelMessageWithSource,
+				Data: &discordgo.InteractionResponseData{
+					Flags:           discordgo.MessageFlagsEphemeral,
+					Content:         "Deleted " + strconv.Itoa(deleted) + " messages from the database.",
+					AllowedMentions: nil,
+				},
+			})
+		}
+	})
+
+	return receiver, nil
+}
+
+func (robo *Robot) onDiscordMessage(ctx context.Context, session *discordgo.Session, event *discordgo.MessageCreate) {
+	// Ignore messages sent by bots
+	if event.Author.Bot {
+		return
+	}
+	log := slog.With(slog.String("trace", event.Message.ID), slog.String("in", event.GuildID))
+
+	// Find configuration for this server
+	ch, _ := robo.channels.Load(event.GuildID)
+	if ch == nil {
+		return
+	}
+
+	// If this is a blocked message we don't want to interact
+	if ch.Block.MatchString(event.Content) {
+		return
+	}
+
+	// Check for the channel being silent. This prevents learning, copypasta,
+	// and random speaking (among other things), which happens to be all the
+	// rest of this function.
+	if s := ch.SilentTime(); event.Timestamp.Before(s) {
+		log.DebugContext(ctx, "channel is silent", slog.Time("until", s))
+		return
+	}
+
+	// TODO: Learn from this user only if not disabled, not sure if applicable to discord
+	// TODO: Meme detector stuff, possibly not necessary in discord?
+
+	// Empty messages could exist due to being a sticker or a file upload
+	if event.Content != "" {
+		robo.discordLearn(ctx, ch.Learn, event)
+	}
+
+	// Now we can check rate limits
+	t := time.Now()
+	r := ch.Rate.ReserveN(t, 1)
+	if d := r.DelayFrom(t); d > 0 {
+		log.InfoContext(ctx, "rate limited",
+			slog.String("action", "speak"),
+			slog.String("delay", d.String()),
+		)
+		r.CancelAt(t)
+		return
+	}
+
+	// Attempt to handle textual commands
+	if strings.HasPrefix(event.Content, session.State.User.Mention()) {
+		// Message starts by mentioning robot
+		trimmed := strings.TrimSpace(strings.TrimPrefix(event.Content, session.State.User.Mention()))
+		parser := regexp.MustCompile(`^(?i:say|generate)\s*(?i:something)?\s*(?i:starting)?\s*(?i:with)?\s+(.*)`)
+		matches := parser.FindStringSubmatch(trimmed)
+		if matches != nil && len(matches) > 1 {
```

Comment:
FYI `len` on a nil slice just gives 0. You don't need to explicitly check for non-nil.